File: frontend/src/app/core/services/project.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { Project, ProjectCreateRequest, ProjectUpdateRequest } from '../models/project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/api/projects';

  list(): Observable<Project[]> {
    return this.http.get<Project[]>(this.baseUrl);
  }

  get(projectId: string): Observable<Project> {
    return this.http.get<Project>(`${this.baseUrl}/${projectId}`);
  }

  create(request: ProjectCreateRequest): Observable<Project> {
    return this.http.post<Project>(this.baseUrl, request);
  }

  update(projectId: string, request: ProjectUpdateRequest): Observable<Project> {
    return this.http.patch<Project>(`${this.baseUrl}/${projectId}`, request);
  }

  delete(projectId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${projectId}`);
  }
}

File: backend/src/main/java/com/testcaseiq/api/project/ProjectService.java
package com.testcaseiq.api.project;

import java.util.List;
import java.util.Locale;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.testcaseiq.api.common.error.ResourceNotFoundException;
import com.testcaseiq.api.domain.model.Project;
import com.testcaseiq.api.domain.repository.ProjectRepository;

@Service
public class ProjectService {

    private final ProjectRepository projectRepository;

    public ProjectService(ProjectRepository projectRepository) {
        this.projectRepository = projectRepository;
    }

    @Transactional
    public ProjectResponse create(ProjectCreateRequest request) {
        Project project = new Project(request.name(), resolveKey(request));
        project.setDescription(request.description());
        return ProjectMapper.toResponse(projectRepository.save(project));
    }

    @Transactional(readOnly = true)
    public List<ProjectResponse> list() {
        return projectRepository.findAllByOrderByCreatedAtDesc()
                .stream()
                .map(ProjectMapper::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public ProjectResponse get(UUID projectId) {
        return ProjectMapper.toResponse(findProject(projectId));
    }

    @Transactional
    public ProjectResponse update(UUID projectId, ProjectUpdateRequest request) {
        Project project = findProject(projectId);
        if (request.name() != null) {
            project.setName(request.name());
        }
        if (request.key() != null) {
            project.setKey(normalizeKey(request.key()));
        }
        if (request.description() != null) {
            project.setDescription(request.description());
        }
        return ProjectMapper.toResponse(project);
    }

    @Transactional
    public void delete(UUID projectId) {
        Project project = findProject(projectId);
        projectRepository.delete(project);
    }

    public Project findProject(UUID projectId) {
        return projectRepository.findById(projectId)
                .orElseThrow(() -> new ResourceNotFoundException("Project not found: " + projectId));
    }

    private String resolveKey(ProjectCreateRequest request) {
        if (request.key() != null && !request.key().isBlank()) {
            return normalizeKey(request.key());
        }
        return normalizeKey(request.name());
    }

    private String normalizeKey(String value) {
        String normalized = value.trim()
                .toLowerCase(Locale.ROOT)
                .replaceAll("[^a-z0-9]+", "-")
                .replaceAll("(^-|-$)", "");
        if (normalized.isBlank()) {
            return "project";
        }
        return normalized.length() > 64 ? normalized.substring(0, 64) : normalized;
    }
}
